### PR TITLE
Monitoring malloc usage

### PR DESCRIPTION
**Pull request by mguetschow**
This is still far from ready to be merged, but before investing more time into it I would like to gather some feedback, see below.

### Contribution description

This is a first attempt of a malloc monitoring implementation that is meant to be used for debugging purposes when working with (libraries that make use of) malloc and friends. Compared to the `MALLOC_TRACING` module already available in RIOT, this is not designed to print the calls to stdout, but to keep track of the amount of memory that is currently dynamically allocated, as well as the all-time max. I was using this while playing with uzlib in #20293.

### Discussion points

- Does such an implementation make sense at all? Or can someone think of another, easier way of monitoring the dynamic memory usage? I tried using [`mallinfo()`](https://linux.die.net/man/3/mallinfo) but that didn't report the all-time max correctly.
- This just hooks into `sys/malloc_thread_safe` which is afaik not used on all boards, and definitely not on `native`. Is there a list somewhere that shows which boards use which malloc implementation? Would it make sense to hook into all of them?
- I would probably implement most of this in a separate module. In that case, do we want `MALLOC_TRACING` to stay around or should it be removed in favor of this implementation? Pinging ~~@benpicco~~ @maribu  here since this was originally his contribution.

### Testing procedure

~~None yet.~~

Test provided in `tests/sys/malloc_monitor`.

### CODE REVIEW

**Comment by maribu** [comment]
Cool! I'm almost certain that some standard c libs (looking avrlibc) do not have mallinfo anyway, so this would be a solution to add this globally.

What would also be quite useful is adding a backtrace to `malloc_state_rm()` when called on an address not traced (e.g. to detect double-free).

> In that case, do we want `MALLOC_TRACING` to stay around or should it be removed in favor of this implementation?

I found that quite useful to trace down a call to `free()` with a wild pointer. If `malloc_state_rm()` where to print the caller address when called on an address not traced, this would work even better for that use case, as it would only print the offending side.

But I guess it can still be useful to resort to highest output verbosity in some cases and the implementation is pretty low-key to maintain, so I guess it could just remain. Maybe just rename that to `malloc_tracing_verbose`, so that the name `malloc_tracing` would be available again?

> This just hooks into `sys/malloc_thread_safe` which is afaik not used on all boards, and definitely not on `native`.

This is correct. What `malloc_thread_safe` is trying to provide is thread-safety on a C lib that does not have reentrant syscalls. E.g. the authors `avrlibc` never imagined someone would be running multiple threads on an AVR, so that will never be thread safe. But also for newlib and picolibc the overhead of enabling reentrant syscalls was prohibitive (at least as a default). Still, corruption and crashes (as concurrent calls to `malloc()` previously caused) was not acceptable and this fixed the most pressing issue with relatively low overhead.

On `native` calls to `malloc()` are thread-safe anyway, so this doesn't wrap them. Enabling `malloc_thread_safe` there anyway (as dependency for `malloc_tracing`) should however not do any harm, other than a bit more overhead.

**Comment by mguetschow** [comment]
Thanks for the feedback! I will then go forward and clean this up.

> What would also be quite useful is adding a backtrace to `malloc_state_rm()` when called on an address not traced (e.g. to detect double-free).

With backtrace you mean basically the same type of print that is used in `malloc_tracing` currently? I can certainly add that.

> But I guess it can still be useful to resort to highest output verbosity in some cases and the implementation is pretty low-key to maintain, so I guess it could just remain. Maybe just rename that to `malloc_tracing_verbose`, so that the name `malloc_tracing` would be available again?

If we want to have the option of high verbosity, I would rather offer a configuration option for the new module and integrate the prints there behind a compile-time flag. I would suppose that a configuration option for a module is nicer for the user than having to deal with two different (pseudo)modules.

I would have proposed `malloc_monitor` as module name, how does that sound? I'd argue that it is _monitoring_ the dynamic heap instead of _tracing_ calls as `malloc_tracing` is currently doing.


> > This just hooks into `sys/malloc_thread_safe` which is afaik not used on all boards, and definitely not on `native`.
> 
> This is correct.

Are there any more boards where `sys/malloc_thread_safe`  is not used? Is enabling it always (also on boards other than `native`) as a dependency of the new module a good option?

**Comment by mguetschow** [comment]
The [murdock failure for native64:llvm](https://ci.riot-os.org/details/434da85ee4344105a13df12f8eea8ca3/tests) is apparently only a single example of all boards with `TOOLCHAIN=llvm` failing, other locally failing boards are `native` and `nrf52840dk`.

It seems the wrapping functions for malloc and friends are not properly handled by clang, which however also means that the thread-safety wrappers are ignored by that toolchain. Interestingly, `tests/sys/malloc_thread_safety` built with `TOOLCHAIN=llvm` does succeed also on `nrf52840dk`, where `malloc_thread_safe` is pulled in by default.

**Comment by mguetschow** [comment]
I have to correct myself, it was a case of the compiler being smarter than the human again :P

Since the memory returned by malloc was never used, all calls to malloc and free where just silently optimized away. Fixed that now by declaring the memory as volatile for the sake of the test.

**Comment by maribu** [comment]
Yeah, we had a test for `calloc()` returning `NULL` when the multiplication was overflowing. LLVM notices that the requested memory was never actually used, so it optimized the allocation out and just claimed that the optimized out allocation succeeded (and `calloc()` returned a non-`NULL` pointer) and the test failed therefore.

**Comment by mguetschow** [comment]
> ./dist/tools/doccheck/check.sh x
  Error: reached end of file while inside a '```' block!

Unfortunately I can't reproduce this locally. What's the doxygen version that's used in the CI?

File: sys/malloc_monitor/doc.txt [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+Note that `malloc_monitor` currently has no notion of threads and will at any point in time report
+the global dynamic memory usage, not the one used by the currently running thread.
+Thread-safety is achieved through usage of @ref sys_malloc_ts, though.
+
+## Example
+
+Imagine you want to investigate the dynamic memory consumption of a certain function `func()`.
+The following snippet could get you started:
+
+```c
```
Comment (maribu) [inline sys/malloc_monitor/doc.txt:35]:
```suggestion
``` C
```

Maybe this pleases Doxygen? (A wild guess only, though)

File: sys/malloc_monitor/malloc_monitor.c [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    size_t high_watermark;
+} malloc_monitor = {
+    .addr = {NULL},
+    .current = 0,
+    .high_watermark = 0,
+};
+
+void malloc_monitor_add(void *ptr, size_t size, uinttxtptr_t pc, char *func_prefix)
+{
+    if (ptr == NULL) return;
```
Comment (maribu) [inline sys/malloc_monitor/malloc_monitor.c:46]:
```suggestion
    if (ptr == NULL) {
        return;
    }
```

The coding convention wants us to always add {} blocks, no matter what. Given the "goto fail" bug, I personally see value in this.

File: sys/malloc_monitor/malloc_monitor.c [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    for (uint8_t i=0; i<CONFIG_MODULE_SYS_MALLOC_MONITOR_SIZE; i++) {
+        if (malloc_monitor.addr[i] == NULL) {
+            malloc_monitor.addr[i] = ptr;
+            malloc_monitor.size[i] = size;
+            malloc_monitor.current += size;
+            if (malloc_monitor.current > malloc_monitor.high_watermark)
+                malloc_monitor.high_watermark = malloc_monitor.current;
```
Comment (maribu) [inline sys/malloc_monitor/malloc_monitor.c:53]:
```suggestion
            if (malloc_monitor.current > malloc_monitor.high_watermark) {
                malloc_monitor.high_watermark = malloc_monitor.current;
            }
```

File: sys/malloc_monitor/doc.txt (hunk 1 above)
Comment (mguetschow) [reply sys/malloc_monitor/doc.txt]: Let's try!

File: sys/malloc_monitor/malloc_monitor.c [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+    }
+    (void)func_prefix;
+    (void)pc;
+    printf("malloc_monitor: maximum number of pointers to be monitored "
+        "(as set by CONFIG_MODULE_SYS_MALLOC_MONITOR_SIZE) exceeded.\n");
+}
+
+void malloc_monitor_rm(void *ptr, uinttxtptr_t pc)
+{
+    if (ptr == NULL) return;
```
Comment (maribu) [inline sys/malloc_monitor/malloc_monitor.c:69]:
```suggestion
    if (ptr == NULL) {
        return;
    }
```

File: sys/malloc_monitor/malloc_monitor.c [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+{
+    if (ptr_old == NULL) {
+        malloc_monitor_add(ptr_new, size_new, pc, "re");
+        return;
+    }
+    if (size_new == 0) {
+        malloc_monitor_rm(ptr_old, pc);
+        return;
+    }
+    if (ptr_new == NULL) return;
```
Comment (maribu) [inline sys/malloc_monitor/malloc_monitor.c:93]:
```suggestion
    if (ptr_new == NULL) {
        return;
    }
```

File: sys/malloc_monitor/malloc_monitor.c [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+    for (uint8_t i=0; i<CONFIG_MODULE_SYS_MALLOC_MONITOR_SIZE; i++) {
+        if (malloc_monitor.addr[i] == ptr_old) {
+            malloc_monitor.addr[i] = ptr_new;
+            size_t size_old = malloc_monitor.size[i];
+            malloc_monitor.size[i] = size_new;
+            if (size_new > size_old) {
+                malloc_monitor.current += size_new - size_old;
+                if (malloc_monitor.current > malloc_monitor.high_watermark)
+                    malloc_monitor.high_watermark = malloc_monitor.current;
```
Comment (maribu) [inline sys/malloc_monitor/malloc_monitor.c:102]:
```suggestion
                if (malloc_monitor.current > malloc_monitor.high_watermark) {
                    malloc_monitor.high_watermark = malloc_monitor.current;
                }
```

File: tests/sys/malloc_monitor/main.c [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+    TEST_ASSERT_CURRENT(0);
+    TEST_ASSERT_WATERMARK(1);
+
+    malloc_monitor_reset_high_watermark();
+
+    TEST_ASSERT_CURRENT(0);
+    TEST_ASSERT_WATERMARK(0);
+}
+
+
```
Comment (maribu) [inline tests/sys/malloc_monitor/main.c:82]:
```suggestion

```

File: sys/malloc_monitor/malloc_monitor.c [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+
+size_t malloc_monitor_get_usage_current(void) {
+    return malloc_monitor.current;
+}
+size_t malloc_monitor_get_usage_high_watermark(void) {
+    return malloc_monitor.high_watermark;
+}
+void malloc_monitor_reset_high_watermark(void) {
+    malloc_monitor.high_watermark = malloc_monitor.current;
+}
```
Comment (maribu) [inline sys/malloc_monitor/malloc_monitor.c:125]:
```suggestion

size_t malloc_monitor_get_usage_current(void) {
    return malloc_monitor.current;
}

size_t malloc_monitor_get_usage_high_watermark(void) {
    return malloc_monitor.high_watermark;
}

void malloc_monitor_reset_high_watermark(void) {
    malloc_monitor.high_watermark = malloc_monitor.current;
}
```

**Review by maribu — COMMENTED** [review]
Looks in general good to me. One open issue I see is synchronization to avoid data races when accessing the monitor data.

On the `malloc()` side, we do already have a `mutex` called. However, a badly timed `malloc_monitor_reset_high_watermark()` would result in data corruption, even on systems where writing to `malloc_monitor.high_watermark` would be an atomic operation.

File: sys/malloc_monitor/doc.txt (hunk 1 above)
Comment (mguetschow) [reply sys/malloc_monitor/doc.txt]: So [the fix](https://github.com/RIOT-OS/RIOT/pull/20363/commits/e54029f9ff12dc2f9906f92b16863993b35fb356) was to use the C documentation comment-style with explicit ` * ` at the beginning of each line.

**Comment by mguetschow** [comment]
Thanks for the review!

> One open issue I see is synchronization to avoid data races when accessing the monitor data.

Very good point o.O, thanks for pointing it out! Addressed with https://github.com/RIOT-OS/RIOT/pull/20363/commits/c5b35283725b79673f7909bbb915388b6ce5c892, please have a look :)

File: sys/malloc_monitor/malloc_monitor.c [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+        if (malloc_monitor.addr[i] == NULL) {
+            malloc_monitor.addr[i] = ptr;
+            malloc_monitor.size[i] = size;
+            malloc_monitor.current += size;
+            if (malloc_monitor.current > malloc_monitor.high_watermark)
+                malloc_monitor.high_watermark = malloc_monitor.current;
+#if CONFIG_MODULE_SYS_MALLOC_MONITOR_VERBOSE
+            printf("%salloc(%" PRIuSIZE ") @ 0x%" PRIxTXTPTR " returned %p\n",
+                func_prefix, size, pc, ptr);
+#endif
```
Comment (gdoffe) [inline sys/malloc_monitor/malloc_monitor.c:70]: I would move this printf before the for loop, as the information displayed do not depend of the number of pointers that can be monitored.

Comment (mguetschow) [reply sys/malloc_monitor/malloc_monitor.c:70]: very true, will change that!

Comment (mguetschow) [reply sys/malloc_monitor/malloc_monitor.c]: Good point, thanks! Adapted in https://github.com/RIOT-OS/RIOT/pull/20363/commits/4f2a3c95dcb57a3d553377b62f29acbaa91076f4

**Comment by mguetschow** [comment]
@maribu @gdoffe maybe you can have another look to include this in the next release :)

File: sys/malloc_monitor/malloc_monitor.c (hunk 9 above)
Comment (gdoffe) [reply sys/malloc_monitor/malloc_monitor.c]: seems good to me !

**Comment by gdoffe** [comment]
> @maribu @gdoffe maybe you can have another look to include this in the next release :)

As a simple user of your work LGTM, good job ! :+1:

**Review by maribu — APPROVED** [review]

**Comment by mguetschow** [comment]
Thank you!

**Pull request merged**